umbra/umbra_x86.c

```diff
@@ -295,11 +300,85 @@ shadow_table_insert_app_to_shadow(void *drcontext,
                                                              sizeof(ptr_int_t),
                                                              disp,
                                                              OPSZ_PTR)));
+}
 #elif defined(ARM)
-    /* FIXME i#1726: port to ARM by adding more XINST_CREATE_ routines */
-#endif
+int
+umbra_num_scratch_regs_for_translation_arch(void)
+{
+    return 2;
 }
 
+/* code sequence:
+ * %reg_idx   = table
+ * %reg_tmp   = (%reg_addr >> 0x10)
+ * %reg_idx  += (%reg_tmp  << 0x02)
+ * %reg_idx   = 0x0(%reg_idx)
+ *
+ * %reg_addr  >>= map->scale
+ * %reg_addr   += %reg_idx
+ */
+static void
+shadow_table_insert_app_to_shadow_arch(void *drcontext, umbra_map_t *map,
+                                       instrlist_t *ilist, instr_t *where,
+                                       reg_id_t reg_addr, reg_id_t *regs)
+{
+    reg_id_t reg_idx = regs[0];
+    reg_id_t reg_tmp = regs[1];
+
+    uint disp_hi = (uint)(map->shadow_table) & 0xffff0000;
+    uint disp_lo = (uint)(map->shadow_table) & 0x0000ffff;
+    disp_hi >>= 0x10;
+
+    /* %reg_idx = table */
+    PRE(ilist, where, INSTR_CREATE_movw(drcontext,
```

Comment:
INSTR_CREATE_load_int() does not load 32-bit ints, so I resorted to using movw and movt. Is there a shorthand for this in the repo already? If not, should there be one?

Reply:
There is instrlist_insert_mov_immed_ptrsz()

Reply:
Whoops, forgot about that method. Done.